src/linear_solver.jl

```diff
@@ -0,0 +1,51 @@
+"""
+    AbstractLinearSolver
+
+All linear solvers used within an `ImplicitFunction` must satisfy this interface.
+
+# Required methods
+
+- `presolve(linear_solver, A, y)`: return a matrix-like object `A` for which it is cheaper to solve several linear systems with different vectors `b` (a typical example would be to perform LU factorization).
+- `solve(linear_solver, A, b)`: return a tuple `(x, stats)` where `x` satisfies `Ax = b` and `stats.solved ∈ {true, false}`.
+"""
+abstract type AbstractLinearSolver end
+
+"""
+    IterativeLinearSolver
+
+An implementation of `AbstractLinearSolver` using `Krylov.gmres`.
+"""
+struct IterativeLinearSolver <: AbstractLinearSolver end
+
+presolve(::IterativeLinearSolver, A, y) = A
+solve(::IterativeLinearSolver, A, b) = gmres(A, b)
+
+"""
+    DirectLinearSolver
+
+An implementation of `AbstractLinearSolver` using the built-in `\` operator.
+"""
+struct DirectLinearSolver <: AbstractLinearSolver end
+
+presolve(::DirectLinearSolver, A, y) = lu(Matrix(A))
+solve(::DirectLinearSolver, A, b) = (A \ b, (; solved=true))
```

Comment:
does not have to be addressed in this PR, but we currently won't fail gracefully is `A` is singular or close to singular

Reply:
adding a try/catch for `SingularException`

Reply:
no try-catch please, it's expensive

Reply:
since it's the `lu` that will fail so you can do
```julia
out = lu(Matrix(A), check = false)
issuccess(out) || throw(SingularException(1))
return out
```

Reply:
perhaps `solved = true` can be replaced by `solved = issuccess(A)`?

Reply:
I was wondering whether we should always error when the solve fails

Reply:
cause if that's the case we can just throw the error in the solver itself, we don't have to check `solved`

and the user can always override it with a custom solver

Reply:
I'm gonna try that, you'll tell me what you think

Reply:
fair enough, I think we shouldn't error, the optimiser should be able to retract if we just return NaN gradient

Reply:
assuming the implicit function is used in some kind of objective function

Reply:
I thought of returning NaNs but they don't exist for every input type

Reply:
At the moment we error anyway, for now I'll just move the error a bit upstream and then we'll see

Reply:
which input type do you have in mind?

Reply:
maybe stuff like BigFloat or Measurements or whatever

Reply:
https://github.com/gdalle/ImplicitDifferentiation.jl/issues/71